src/remote.c

```diff
@@ -670,21 +670,46 @@ int git_remote_set_pushurl(git_repository *repo, const char *remote, const char*
 	return set_url(repo, remote, CONFIG_PUSHURL_FMT, url);
 }
 
-const char* git_remote__urlfordirection(git_remote *remote, int direction)
+static int resolve_url(git_buf *resolved_url, const char *url, int direction, const git_remote_callbacks *callbacks)
 {
-	assert(remote);
+	int status;
 
-	assert(direction == GIT_DIRECTION_FETCH || direction == GIT_DIRECTION_PUSH);
+	if (callbacks && callbacks->resolve_url) {
+		git_buf_clear(resolved_url);
+		status = callbacks->resolve_url(resolved_url, url, direction, callbacks->payload);
+		if (status != GIT_PASSTHROUGH) {
+			git_buf_sanitize(resolved_url);
+			return status;
+		}
+	}
+
+	git_buf_sets(resolved_url, url);
+
+	return 0;
+}
+
+int git_remote__urlfordirection(git_buf *url_out, struct git_remote *remote, int direction, const git_remote_callbacks *callbacks)
+{
+	const char *url = NULL;
+
+	assert(remote);
 
 	if (direction == GIT_DIRECTION_FETCH) {
-		return remote->url;
+		url = remote->url;
+	} else if (direction == GIT_DIRECTION_PUSH) {
+		url = remote->pushurl ? remote->pushurl : remote->url;
+	} else {
+		assert(direction == GIT_DIRECTION_FETCH || direction == GIT_DIRECTION_PUSH);
```

Comment:
Style nit: the assert should probably be moved to the top so that preconditions to that function are easily spottable.

Reply:
I was evading Xcode's warnings (for the one time it warns about new things, it's a complete crazy one 😭). The analyzer saw the access to an uninitialized `url` if that assert is hoisted above, because (I think) the `if/else` makes it reconsider its analysis that it's only 2 values.

Frankly, it's the first time I'm seeing this warning, and it's… strange. And gone. Oh Xcode.

Reply:
It cannot be uninitialized as you're initializing it to `NULL` above. Huh. Maybe you've added the `NULL` initializer at a later point?

Reply:
That ought to be it (this has just been fixed-up and rebased).